src/WidgetBase.ts

```diff
@@ -0,0 +1,525 @@
+import { VNode, VNodeProperties } from '@dojo/interfaces/vdom';
+import { assign } from '@dojo/core/lang';
+import WeakMap from '@dojo/shim/WeakMap';
+import Promise from '@dojo/shim/Promise';
+import Map from '@dojo/shim/Map';
+import Set from '@dojo/shim/Set';
+import { EventTypedObject } from '@dojo/interfaces/core';
+import { Evented } from './bases/Evented';
+import { v, registry, isWNode } from './d';
+import FactoryRegistry from './FactoryRegistry';
+
+/**
+ * Base widget properties
+ */
+export interface WidgetProperties {
+
+	/**
+	 * id for a widget
+	 */
+	id?: string;
+
+	/**
+	 * The key for a widget. Used to differentiate uniquely identify child widgets for
+	 * rendering and instance management
+	 */
+	key?: string;
+
+	/**
+	 * The scope to bind all function properties
+	 */
+	bind?: any;
+}
+
+/**
+ * Wrapper for v
+ */
+export interface HNode {
+	/**
+	 * Array of processed VNode children.
+	 */
+	vNodes?: (string | VNode | null)[];
+	/**
+	 * Specified children
+	 */
+	children: (DNode | string)[];
+
+	/**
+	 * render function that wraps returns VNode
+	 */
+	render<T>(options?: { bind?: T }): VNode;
+
+	/**
+	 * The properties used to create the VNode
+	 */
+	properties: VNodeProperties;
+
+	/**
+	 * The type of node
+	 */
+	type: symbol;
+}
+
+/**
+ * Wrapper for `w`
+ */
+export interface WNode {
+	/**
+	 * Factory to create a widget
+	 */
+	factory: WidgetConstructor | string;
+
+	/**
+	 * Options used to create factory a widget
+	 */
+	properties: WidgetProperties;
+
+	/**
+	 * DNode children
+	 */
+	children: DNode[];
+
+	/**
+	 * The type of node
+	 */
+	type: symbol;
+}
+
+/**
+ * union type for all possible return types from render
+ */
+export type DNode = HNode | WNode | string | null;
+
+/**
+ * the event emitted on properties:changed
+ */
+export interface PropertiesChangeEvent<T, P extends WidgetProperties> extends EventTypedObject<'properties:changed'> {
+	/**
+	 * the full set of properties
+	 */
+	properties: P;
+	/**
+	 * the changed properties between setProperty calls
+	 */
+	changedPropertyKeys: string[];
+	/**
+	 * the target (this)
+	 */
+	target: T;
+}
+
+/**
+ * Propeerty Change record for specific property diff functions
+ */
+export interface PropertyChangeRecord {
+	changed: boolean;
+	value: any;
+}
+
+/**
+ * Properties changed record, return for diffProperties
+ */
+export interface PropertiesChangeRecord<P extends WidgetProperties> {
+	changedKeys: string[];
+	properties: P;
+}
+
+/**
+ * WidgetBase constructor type
+ */
+export type WidgetConstructor = new (...args: any[]) => WidgetBase<WidgetProperties>;
```

Comment:
I think that absolutely makes sense @maier49, do you mind if we fix the up in a follow up PR?